Feature/user groups

apurvann opened the pull request.

apurvann committed:
user groups with rules engine. wip

apurvann committed:
first working set of rest apis and frontend js

apurvann committed:
first version of functional groups/blueprints/instances

group joining code generated randomly

apurvann committed:
blueprint templates and blueprints working with groups

apurvann committed:
import/export functional for blueprints and blueprint templates

apurvann committed:
group users fix and rules expression builder

apurvann committed:
api tests for user groups with blueprints / blueprint templates

python3.4 builds fixes

tests for group joining banned user, import/export blueprint templates and blueprint/templates full_config

apurvann committed:
public groups

system group creation checks

apurvann committed:
Groups misc. tuning

system groups test fix

apurva3000 commented:
Do not merge it yet.

tourunen commented on pouta_blueprints/models.py:
```diff
@@ -144,6 +145,49 @@ def verify_auth_token(token, app_secret):
+        return hash(self.email)
+
+
+group_user = db.Table('groups_users', db.Column('group_id', db.String(32), db.ForeignKey('groups.id')), db.Column('user_id', db.String(32), db.ForeignKey('users.id')), db.PrimaryKeyConstraint('group_id', 'user_id'))
```
break these long lines up

tourunen reviewed (commented):
Looks pretty good

apurvann committed:
group owners and manager roles defined

migrations

apurvann committed:
group exit feature with dynamic user config

group joining enhancements

apurvann committed:
some blueprint fields now hybrid to reflect template changes

rules engine refactoring and group deletion for admin ui

apurvann committed:
association objects for owner and manager roles. ui changes for group listing and joining

group and account ui view for admins should be different. system groups and sorting groups for admins

apurvann committed:
fine tuning the groups/templates feature

missing tests added

docstrings added

Jyrsa commented on pouta_blueprints/forms.py:
```diff
@@ -26,26 +28,48 @@ class UserForm(ModelForm):
+    template_name = StringField('template_name', validators=[DataRequired()])
+    group_name = StringField('group_name', validators=[DataRequired()])
+
+# class BlueprintImportForm(Form):
```
should this still be here?

Jyrsa commented on pouta_blueprints/models.py:
```diff
@@ -237,17 +294,72 @@ def model(self, value):
+    __tablename__ = 'blueprint_templates'
+    id = db.Column(db.String(32), primary_key=True)
+    name = db.Column(db.String(MAX_NAME_LENGTH))
+    _config = db.Column('config', db.Text)
```
as a whole I'm not a big fan of this style of property storing in databases but it seems to be the way things have been done in this project. Now is not the time to change it but suffice to say that I'd like to at some point.

Jyrsa commented on pouta_blueprints/rules.py:
```diff
@@ -0,0 +1,110 @@
+from sqlalchemy.orm import load_only
+from sqlalchemy.sql.expression import true
+import itertools
+# import logging
```
commented code

Jyrsa commented on pouta_blueprints/rules.py:
```diff
@@ -0,0 +1,110 @@
+
+
+def apply_rules_blueprint_templates(user, args=None):
+
```
This empty line is just screaming for a simple docstring. Doesn't have to be more than one sentence but that sentence reduces cognitive load of reader.

Jyrsa commented on pouta_blueprints/rules.py:
```diff
@@ -0,0 +1,110 @@
+
+
+def apply_rules_blueprints(user, args=None):
+
```
see above

Jyrsa commented on pouta_blueprints/rules.py:
```diff
@@ -0,0 +1,110 @@
+
+
+def apply_rules_export_blueprints(user):
+    q = Blueprint.query
```
maybe add a whitespace here to be consistent?

Jyrsa replied:
Or a docstring as well. This one is almost quickly legible but even then reasoning about DB api code can be something of a cognitive load.

Jyrsa commented on pouta_blueprints/rules.py:
```diff
@@ -0,0 +1,110 @@
+
+
+def get_manager_group_ids(user):
+    # the result shall contain the owners of the groups too as they are managers by default
```
What exactly does this function do? Get the managers for the groups a user is a member in?

Jyrsa replied:
In retrospect probably, but again a good idea to use plain language.

Jyrsa commented on pouta_blueprints/static/js/main.js:
```diff
@@ -2,7 +2,7 @@
 'use strict';
 
-var app = angular.module('PBApp', ['ngRoute', 'restangular', 'LocalStorageModule', 'validation.match', 'angularFileUpload', 'schemaForm', 'ui.bootstrap', 'angular-loading-bar']);
+var app = angular.module('PBApp', ['ngRoute', 'restangular', 'LocalStorageModule', 'validation.match', 'angularFileUpload', 'schemaForm', 'ui.bootstrap', 'angular-loading-bar', 'angularjs-dropdown-multiselect']);
```
maybe split over several lines?

Jyrsa commented on pouta_blueprints/static/js/services/AuthService.js:
```diff
@@ -57,6 +59,25 @@ app.factory('AuthService', ['$q', 'localStorageService', 'Session', 'Restangular
             return false;
         },
 
+        isGroupOwnerOrAdmin : function() {
```
trusting the localstorage makes it possible for the user to ruin their user experience by trying to hack things but i suppose this is acceptable as long as the checks in the views are valid.

Jyrsa commented on pouta_blueprints/utils.py:
```diff
@@ -121,3 +132,47 @@ def parse_port_range(port_range):
+
+
+def get_full_blueprint_config(blueprint):
+
```
code is quite clear on what happens but the word "full" in function name is ambiguous. A quick docstring would help with this.

Jyrsa commented on pouta_blueprints/utils.py:
```diff
@@ -121,3 +132,47 @@ def parse_port_range(port_range):
+    full_config = get_full_blueprint_config(blueprint)
+
+    if field_name == 'preallocated_credits':
+        preallocated_credits = False  # Default value
```
should we maybe store the defaults in a dict or a namespace so that they're all in one place?

Jyrsa replied:
commons.py might be a good place.

Jyrsa commented on pouta_blueprints/views/groups.py:
```diff
@@ -0,0 +1,293 @@
+
+
+def group_users_add(group, user_config, owner):
+    # Generate a 'set' of Group Managers
```
A docstring?

Jyrsa commented on pouta_blueprints/views/groups.py:
```diff
@@ -0,0 +1,293 @@
+
+
+def generate_user_config(group):
+
```
inconsistency in whitespace between def line and code.

Jyrsa commented on pouta_blueprints/views/instances.py:
```diff
@@ -74,15 +75,8 @@ class InstanceList(restful.Resource):
     def get(self):
```
This function calls a lot of helper functions so it might be helpful for the next reader to have a simplified human-readable explanation of what instances a user or admin is allowed to see.

Jyrsa commented on pouta_blueprints/views/users.py:
```diff
@@ -39,7 +39,15 @@ def post(self):
+                .order_by(User.is_group_owner.desc())
+                .order_by(User.is_active)
+                .order_by(User.is_blocked)
+                .order_by(User.is_blocked)
```
is this something incredibly genius that i don't understand or a duplicate line?

Jyrsa commented on pouta_blueprints/views/users.py:
```diff
@@ -202,3 +210,27 @@ def put(self, user_id):
+        args = self.parser.parse_args()
+        make_group_owner = args.make_group_owner
+
+        user = User.query.filter_by(id=user_id).first()
```
Doesn't flask also have a get() for these occasions? Or do you just want to avoid catching the error a get would throw if the key is not found?

Jyrsa reviewed (commented):
Mostly minor comments, whitespace typos and docstring requests.
